File: 20.valid-parentheses.go
package main

/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

// @lc code=start

func isValid(s string) bool {
	stack := make([]rune, 0)

	brackets := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	for _, char := range s {
		if closeStr, ok := brackets[char]; ok {
			if len(stack) == 0 || stack[len(stack)-1] != closeStr {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, char)
		}
	}

	return len(stack) == 0
}

// @lc code=end

func _main() {
	result := isValid("{{[]}}")
	println(result)
}

File: 20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::new();
    
        let mut map = std::collections::HashMap::new();
        map.insert(')', '(');
        map.insert('}', '{');
        map.insert(']', '[');
    
        for char in s.chars() {
            match map.get(&char) {
                Some(char) => {
                    if stack.len() == 0 || stack.last().unwrap() != char {
                        return false;
                    }
    
                    let _ = stack.pop();
                }
                None => {
                    stack.push(char);
                }
            }
        }
    
        stack.len() == 0
    }
}
// @lc code=end

